src/object.rs

```diff
@@ -171,56 +202,323 @@ impl InMemNiftiObject {
     /// a header.
-    pub fn new_from_stream<R: Read>(&self, mut source: R) -> Result<InMemNiftiObject> {
-        let header = NiftiHeader::from_stream(&mut source)?;
+    pub fn from_reader(mut source: R) -> Result<Self> {
+        let header = NiftiHeader::from_reader(&mut source)?;
         if &header.magic == MAGIC_CODE_NI1 {
             return Err(NiftiError::NoVolumeData);
         }
         let len = header.vox_offset as usize;
         let len = if len < 352 { 0 } else { len - 352 };
-        let extender = Extender::from_stream(&mut source)?;
+        let extender = Extender::from_reader(&mut source)?;
         
         let ext = {
             let source = ByteOrdered::runtime(&mut source, header.endianness);
-            ExtensionSequence::from_stream(extender, source, len)?
+            ExtensionSequence::from_reader(extender, source, len)?
         };
 
-        let volume = InMemNiftiVolume::from_stream(source, &header)?;
+        let volume = StreamedNiftiVolume::from_reader(source, &header)?;
 
-        Ok(InMemNiftiObject {
+        Ok(StreamedNiftiObject {
             header,
             extensions: ext,
             volume,
         })
     }
+
+    /// Read a NIFTI volume, and extensions, from a stream of data. The header,
+    /// extender code and expected byte order of the volume's data must be
+    /// known in advance.
+    fn from_reader_with_extensions(
+        mut source: R,
+        header: &NiftiHeader,
+        extender: Extender,
+    ) -> Result<(StreamedNiftiVolume<R>, ExtensionSequence)>
+    where
+        R: Read,
+    {
+        // fetch extensions
+        let len = header.vox_offset as usize;
+        let len = if len < 352 { 0 } else { len - 352 };
+
+        let ext = {
+            let source = ByteOrdered::runtime(&mut source, header.endianness);
+            ExtensionSequence::from_reader::<_, _>(extender, source, len)?
+        };
+
+        // fetch volume (rest of file)
+        Ok((StreamedNiftiVolume::from_reader(source, &header)?, ext))
+    }
 }
 
-impl NiftiObject for InMemNiftiObject {
-    type Volume = InMemNiftiVolume;
+impl StreamedNiftiObject<Box<dyn Read>> {
+    /// Retrieve the NIfTI object and prepare the volume for streamed reading.
+    /// The given file system path is used as reference.
+    /// If the file only contains the header, this method will
+    /// look for the corresponding file with the extension ".img",
+    /// or ".img.gz" if the former wasn't found.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use nifti::{NiftiObject, StreamedNiftiObject};
+    /// # use nifti::error::Result;
+    ///
+    /// # fn run() -> Result<()> {
+    /// let obj = StreamedNiftiObject::from_file("minimal.nii.gz")?;
+    /// 
+    /// let volume = obj.into_volume();
+    /// for slice in volume {
+    ///     // manipulate slice here
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
```

Comment:
Is this supposed to be indented like a normal function would be?

Reply:
Since the outer function constructs are invisible in the doctests, it is preferable not to indent the body. What could be done now is take advantage of `?` in doctests (since 1.34) and discard the wrapper boilerplate entirely. I've done so in a few other places, one might as well do the same for new functions.